File: core-v3/src/main/java/com/adyen/checkout/core/log/LogUtil.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 11/2/2019.
 */

package com.adyen.checkout.core.log;

import android.os.Build;
import androidx.annotation.NonNull;

import com.adyen.checkout.core.exception.NoConstructorException;
import com.adyen.checkout.core.util.KotlinBase;

/**
 * Utility class with methods related to logs.
 */
public final class LogUtil {

    private static final String CHECKOUT_LOG_PREFIX = "CO.";
    private static final String CLASS_NOT_FOUND = "?Unknown?";

    private static final int MAX_TAG_SIZE = 23;

    static {
        KotlinBase.Companion.log();
    }

    /**
     * Get the TAG to be used for logging inside Checkout classes.
     *
     * @return A String to be used as TAG.
     */
    @NonNull
    public static String getTag() {
        return getTag(CHECKOUT_LOG_PREFIX);
    }

    /**
     * Get the TAG to be used for logging by the calling class.
     *
     * @return A String to be used as TAG with the format "Prefix.ClassName"
     */
    //This could be used by merchants if they want to.
    @SuppressWarnings("WeakerAccess")
    @NonNull
    public static String getTag(@NonNull String prefix) {
        final String callerClass = getSimplifiedCallerClassName();
        String tag = prefix + callerClass;

        // Log tags have a size limitation on API lvl 23 and before
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.M && tag.length() > MAX_TAG_SIZE) {
            tag = tag.substring(0, MAX_TAG_SIZE);
        }

        return tag;
    }

    @NonNull
    private static String getSimplifiedCallerClassName() {
        final String className = getCallerClassName();
        return simplifyClassName(className);
    }

    @NonNull
    private static String getCallerClassName() {
        final StackTraceElement[] stElements = Thread.currentThread().getStackTrace();
        for (int i = 1; i < stElements.length; i++) {
            final StackTraceElement ste = stElements[i];
            final String callerClass = ste.getClassName();
            if (!callerClass.equals(LogUtil.class.getName()) && callerClass.indexOf("java.lang.Thread") != 0) {
                return callerClass;
            }
        }
        return CLASS_NOT_FOUND;
    }

    @NonNull
    private static String simplifyClassName(@NonNull String className) {
        final String[] packageSplit = className.split("\\.");
        if (packageSplit.length == 0) {
            return className;
        }
        return packageSplit[packageSplit.length - 1];
    }

    private LogUtil() {
        throw new NoConstructorException();
    }

}

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 17/12/2020.
 */
package com.adyen.checkout.core.log

import android.os.Build

/**
 * Utility class with methods related to logs.
 */
object LogUtil {

    private const val CHECKOUT_LOG_PREFIX = "CO."
    private const val CLASS_NOT_FOUND = "?Unknown?"
    private const val MAX_TAG_SIZE = 23

    /**
     * Get the TAG to be used for logging inside Checkout classes.
     *
     * @return A String to be used as TAG.
     */
    @JvmStatic
    fun getTag(): String = getTag(CHECKOUT_LOG_PREFIX)

    /**
     * Get the TAG to be used for logging by the calling class.
     *
     * @return A String to be used as TAG with the format "Prefix.ClassName"
     */
    // This could be used by merchants if they want to.
    @JvmStatic
    fun getTag(prefix: String): String {
        val callerClass = simplifiedCallerClassName
        var tag = prefix + callerClass

        // Log tags have a size limitation on API lvl 23 and before
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.M && tag.length > MAX_TAG_SIZE) {
            tag = tag.substring(0, MAX_TAG_SIZE)
        }
        return tag
    }

    private val simplifiedCallerClassName: String
        get() {
            val className = callerClassName
            return simplifyClassName(className)
        }

    private val callerClassName: String
        get() {
            val stElements = Thread.currentThread().stackTrace
            for (ste in stElements.drop(1)) {
                val callerClass = ste.className
                if (callerClass != LogUtil::class.java.name && callerClass.indexOf("java.lang.Thread") != 0) {
                    return callerClass
                }
            }
            return CLASS_NOT_FOUND
        }

    private fun simplifyClassName(className: String): String {
        val packageSplit = className.split(".").toTypedArray()
        return if (packageSplit.isEmpty()) {
            className
        } else packageSplit[packageSplit.size - 1]
    }
}
